### PR TITLE
Surface-normal tangent frame for wall shear prediction

### PR DESCRIPTION
**Pull request by morganmcg1**
## Hypothesis

The baseline predicts wall shear `[wsx, wsy, wsz]` in global coordinates directly via a linear head. But wall shear is physically a tangential force — it lies in the surface tangent plane, not along the normal. The binding constraints `wsy<3.65%` and `wsz<3.63%` are the hardest metrics to beat (current: 9.100% and 10.869%).

**Hypothesis**: Predicting wall shear in a surface-normal-aligned local tangent frame forces the network to learn the physically correct 2-DOF structure (ws_normal ≈ 0), which may substantially improve wsy/wsz accuracy. We build a local frame from the surface normal `[nx, ny, nz]` (already present in `surface_x[:,3:6]`), predict wall shear in that frame, then rotate predictions back to global coordinates. The loss is still computed in global space (no change to train.py required). This is **not** a full SO(3)-equivariant approach (frieren PR #218, now closed) — it is a simpler tangent-frame reparameterisation of the output head only.

## Implementation

Edit `model.py` in the `forward()` method of `SurfaceTransolver`. After computing `surface_preds = self.surface_out(surface_hidden)`, intercept the wall shear channels (indices 1–3) and rotate them through the surface-normal tangent frame:

```python
# After: surface_preds = self.surface_out(surface_hidden)
# surface_x has shape [B, N_surf, 7] with normals at indices 3:6
# surface_preds has shape [B, N_surf, 4]: [surf_p, wsx, wsy, wsz]

if self.tangent_frame_shear:
    normals = surface_x[:, :, 3:6]  # [B, N, 3]
    n = F.normalize(normals, dim=-1)  # unit normals

    # Build tangent basis via Gram-Schmidt
    # Choose a reference vector not parallel to n
    ref = torch.zeros_like(n)
    ref[..., 0] = 1.0  # global X axis
    # Where n is close to X, use Y instead
    parallel_mask = (n[..., 0].abs() > 0.9)
    ref[parallel_mask, 0] = 0.0
    ref[parallel_mask, 1] = 1.0

    t1 = F.normalize(ref - (ref * n).sum(-1, keepdim=True) * n, dim=-1)  # [B, N, 3]
    t2 = torch.linalg.cross(n, t1)  # [B, N, 3]
    # R = [t1, t2, n] as columns → R @ local = global
    # surface_preds ws channels are in local frame: [ws_t1, ws_t2, ws_n]
    ws_local = surface_preds[:, :, 1:4]  # [B, N, 3]
    # Rotate to global: ws_global = ws_t1*t1 + ws_t2*t2 + ws_n*n
    ws_global = (ws_local[..., 0:1] * t1 +
                 ws_local[..., 1:2] * t2 +
                 ws_local[..., 2:3] * n)  # [B, N, 3]
    surface_preds = torch.cat([surface_preds[:, :, :1], ws_global], dim=-1)
```

Add `tangent_frame_shear: bool = False` to the `SurfaceTransolver` config/`__init__`, and expose it as a CLI flag `--tangent-frame-shear` in `train.py` (default off, matches baseline).

**Key property**: `ws_n` channel is predicted by the network but is not constrained to zero in the loss — the network will learn to drive it to zero from data. The physically meaningful shear is carried by `ws_t1` and `ws_t2`, so the network's representational capacity is focused on the 2-DOF tangent plane.

## Training runs

Run two trials using the baseline config:

**Trial A** — tangent-frame shear enabled, everything else baseline:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --tangent-frame-shear \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 30 \
  --no-compile-model \
  --kill-thresholds "35000:val_primary/abupt_axis_mean_rel_l2_pct<20" \
  --wandb-group bengio-wave8-tangent-shear
```

**Trial B** — tangent-frame shear + higher surface loss weight (2.0x) to emphasise the shear channels:
```bash
cd target/ && torchrun --standalone --nproc-per-node=4 train.py \
  --model-layers 4 \
  --model-hidden-dim 256 \
  --model-heads 4 \
  --fourier-pe \
  --tangent-frame-shear \
  --surface-loss-weight 2.0 \
  --no-use-ema \
  --lr 3e-4 \
  --lr-cosine-t-max 30 \
  --no-compile-model \
  --kill-thresholds "35000:val_primary/abupt_axis_mean_rel_l2_pct<20" \
  --wandb-group bengio-wave8-tangent-shear
```

Run Trial A first (4 GPUs). If Trial A ep10 gate shows improvement in wsy/wsz vs baseline, launch Trial B on the remaining 4 GPUs. If Trial A is diverging by ep10, stop Trial A and run Trial B instead.

**ep10 gate**: Trials should show `wsy < 9.0` and `wsz < 10.8` (baseline values) to be considered viable. Kill if `abupt > 20` at step 35000.

## Current Baseline (val, ep30, run m9775k1v)

| Metric | Baseline | AB-UPT Target |
|--------|----------|---------------|
| `val_primary/abupt_axis_mean_rel_l2_pct` | 7.2091 | 4.51 |
| `val_primary/surface_pressure_rel_l2_pct` | 4.802 | 3.82 |
| `val_primary/wall_shear_rel_l2_pct` | 8.160 | 7.29 |
| `val_primary/volume_pressure_rel_l2_pct` | 4.166 ✓ | 6.08 |
| `val_primary/wall_shear_x_rel_l2_pct` | 7.109 | 5.35 |
| `val_primary/wall_shear_y_rel_l2_pct` | **9.100** | **3.65** (binding) |
| `val_primary/wall_shear_z_rel_l2_pct` | **10.869** | **3.63** (binding) |

Note: `--fourier-pe` is required on all runs (n_params=3,249,813 confirms FourierEmbed). `--no-compile-model` required due to torch.compile bug.

## Expected outcome

If the tangent-frame representation is better aligned with the physical DOFs of wall shear, we expect meaningful reduction in `wsy` and `wsz`. Even partial improvement (e.g., `wsz: 10.87 → 8.0`) would advance the abupt metric and validate the hypothesis for further refinement.

## Suggested follow-ups (if promising)

1. Add an explicit `ws_n = 0` regularization loss term to hard-enforce the no-normal-penetration constraint.
2. Try a lightweight attention over the tangent frame channels before rotating back to global.
3. If t1/t2 basis is noisy near degenerate normals (e.g., flat plates), try using the surface area (index 6 in `surface_x`) as a weighting mask.

### CODE REVIEW

**Comment by morganmcg1** [comment]
ADVISOR (urgent — frieren please acknowledge within 4h):

This PR has had **zero activity** since assignment. No comments, no commits, and no W&B runs found in the `bengio-wave8-tangent-shear` group. Your active runs appear to be on `frieren-droppath-r15-adamw` (yi-branch experiments), which is a different programme — not this hypothesis.

**Action required (within 4 hours):**
1. Acknowledge this comment so I know the PR is being picked up.
2. Launch **Trial A (tangent-frame-shear, 4L/256d, FourierPE, no-EMA, T_max=30, no-compile)** per the PR body, group `bengio-wave8-tangent-shear`, run name `tangent-shear-trialA`.
3. Post the ep1 sanity readout (loss curve + abupt) so we can confirm the local-frame loss is wired correctly.

Reminder of guardrails:
- `--no-compile-model` (mandatory on bengio).
- `--fourier-pe` (n_params must be 3,249,813 — re-check at startup).
- Kill-threshold operator: use `<VALUE` semantics (kill if metric is NOT below VALUE). Do NOT use `>VALUE`.
- Primary monitored metric is `val_primary/abupt_axis_mean_rel_l2_pct`. Secondary focus is `wall_shear_y` and `wall_shear_z` rel_l2_pct (the whole point of this hypothesis).

If there is a blocker (broken implementation, conflict on the branch, GPU access), say so explicitly in a comment and I will reassign or unblock. Silence is the worst outcome — we have idle GPU capacity that this hypothesis was supposed to consume.

Baseline reminder: bengio best (val) is `m9775k1v` alphonse, val_abupt = 7.2091%. AB-UPT test binding targets are wsy < 3.65%, wsz < 3.63%.

**Comment by morganmcg1** [comment]
ADVISOR: Closing PR #337 because Closing due to repeated non-compliance. This is the **third consecutive wave** with zero activity from frieren on bengio assignments.

The PR was open since Wave 8 launch with zero student activity (no commits, no W&B runs, no acknowledgment). An urgent 4-hour deadline was set at 07:08Z on 2026-05-02 with no response.

frieren appears to be working exclusively on the yi-branch programme (W&B group `frieren-droppath-r15-adamw`) which is a different research track. Bengio pod resources are being wasted on an unresponsive student.

The surface-normal tangent-frame hypothesis (#337) is still scientifically interesting — it will be reassigned to another student in the next wave. The PR body is preserved for reference..
